edb/graphql/translator.py

```diff
@@ -1510,24 +1547,100 @@ def value_node_from_pyvalue(val: Any):
         raise ValueError(f'unexpected constant type: {type(val)!r}')
 
 
-def translate(gqlcore: gt.GQLCoreSchema, query, *,
-              operation_name=None, variables=None):
+def parse_text(query: str) -> graphql.Document:
     try:
-        document_ast = graphql.parse(query)
+        return graphql.parse(query)
     except graphql.GraphQLError as err:
         err_loc = (err.locations[0].line,
                    err.locations[0].column)
         raise g_errors.GraphQLCoreError(err.message, loc=err_loc) from None
 
+
+class TokenLexer(graphql.language.lexer.Lexer):
+
+    def __init__(self, source, tokens, eof_pos):
+        self.__tokens = tokens
+        self.__index = 0
+        self.__eof_pos = eof_pos
+        self.source = source
+        kind, start, end, line, col, body = self.__tokens[0]
+        self.token = gql_lexer.Token(kind, start, end, line, col, None, body)
+
+    def advance(self) -> gql_lexer.Token:
+        self.last_token = self.token
+        token = self.token = self.lookahead()
+        self.__index += 1
+        return token
+
+    def lookahead(self) -> gql_lexer.Token:
+        token = self.token
+        if token.kind != gql_lexer.TokenKind.EOF:
+            if token.next:
+                return self.token.next
+            kind, start, end, line, col, body = self.__tokens[self.__index + 1]
+            token.next = gql_lexer.Token(
+                kind, start, end, line, col, token, body)
+            return token.next
+        else:
+            return token
+
+
+def parse_tokens(
+    text: str,
+    tokens: List[Tuple[gql_lexer.TokenKind, int, int, int, int, str]]
+) -> graphql.Document:
+    try:
+        src = graphql.Source(text)
+        parser = graphql.language.parser.Parser(src)
+        parser._lexer = TokenLexer(src, tokens, len(text))
+        return parser.parse_document()
+    except graphql.GraphQLError as err:
+        err_loc = (err.locations[0].line,
+                   err.locations[0].column)
+        raise g_errors.GraphQLCoreError(err.message, loc=err_loc) from None
+
+
+def convert_errors(
+    errs: List[gql_error.GraphQLError], *,
+    substitutions: Optional[Dict[str, Tuple[str, int, int]]],
+) -> List[gql_error.GraphQLErrors]:
+    result = []
+    for err in errs:
+        m = REWRITE_TYPE_ERROR.match(err.message)
+        if not m:
+            # we allow convefsion from Int to Float, and that is allowed by
+            # graphql spec. It's unclear why graphql-core chokes on this
+            if INT_FLOAT_ERROR.match(err.message):
```

Comment:
This is fragile. Do we have a functional test for this specific hack to track regressions?

Reply:
This is caught by `test_graphql_functional_variables_03` and `test_graphql_functional_variables_10`.